pkg/didcomm/protocol/issuecredential/service.go

```diff
@@ -0,0 +1,417 @@
+	CredentialPreviewMsgType = Spec + "credential-preview"
+)
+
+const stateNameKey = "state_name_"
+
+var logger = log.New("aries-framework/issuecredential/service")
+
+// customError is a wrapper to determine custom error against internal error
+type customError struct{ error }
+
+// transitionalPayload keeps payload needed for Continue function to proceed with the action
+type transitionalPayload struct {
+	// protocol state machine identifier
+	PIID      string
+	StateName string
+	Msg       service.DIDCommMsgMap
+	MyDID     string
+	TheirDID  string
+}
+
+// metaData type to store data for internal usage
+type metaData struct {
+	transitionalPayload
+	state    state
+	msgClone service.DIDCommMsg
+	inbound  bool
+	// err is used to determine whether callback was stopped
+	// e.g the user received an action event and executes Stop(err) function
+	// in that case `err` is equal to `err` which was passing to Stop function
+	err error
+}
+
+// Provider contains dependencies for the protocol and is typically created by using aries.Context()
+type Provider interface {
+	Messenger() service.Messenger
+	StorageProvider() storage.Provider
+}
+
+// Service for the issuecredential protocol
+type Service struct {
+	service.Action
+	service.Message
+	store     storage.Store
+	callbacks chan *metaData
+	messenger service.Messenger
+}
+
+// New returns the issuecredential service
+func New(p Provider) (*Service, error) {
+	store, err := p.StorageProvider().OpenStore(Name)
+	if err != nil {
+		return nil, err
+	}
+
+	svc := &Service{
+		messenger: p.Messenger(),
+		store:     store,
+		callbacks: make(chan *metaData),
+	}
+
+	// start the listener
+	go svc.startInternalListener()
+
+	return svc, nil
+}
+
+// HandleInbound handles inbound message (issuecredential protocol)
+func (s *Service) HandleInbound(msg service.DIDCommMsg, myDID, theirDID string) (string, error) {
+	aEvent := s.ActionEvent()
+
+	// throw error if there is no action event registered for inbound messages
+	if aEvent == nil {
+		return "", errors.New("no clients are registered to handle the message")
+	}
+
+	md, err := s.doHandle(msg, false)
+	if err != nil {
+		return "", fmt.Errorf("doHandle: %w", err)
+	}
+
+	// sets inbound payload
+	md.inbound = true
+	md.MyDID = myDID
+	md.TheirDID = theirDID
+
+	// trigger action event based on message type for inbound messages
+	if canTriggerActionEvents(msg) {
+		aEvent <- s.newDIDCommActionMsg(md)
+
+		return "", nil
+	}
+
+	// if no action event is triggered, continue the execution
+	return "", s.handle(md)
+}
+
+// HandleOutbound handles outbound message (issuecredential protocol)
+func (s *Service) HandleOutbound(msg service.DIDCommMsg, myDID, theirDID string) (string, error) {
+	md, err := s.doHandle(msg, true)
+	if err != nil {
+		return "", fmt.Errorf("doHandle: %w", err)
+	}
+
+	// sets outbound payload
+	md.MyDID = myDID
+	md.TheirDID = theirDID
+
+	return "", s.handle(md)
```

Comment:
what is the `string` return arg for?

Reply:
It is because of the service interface. For did exchange we are returning `connectionID`. For the `introduce` and `issue-credential` protocols we do not need this. 

NOTE: I am going to refactor it later. In my previous PR, I have added this for HandleOutbound. But it seems to be redundant.